### ISSUE TITLE
The communication between the ESP32C3 development board and the board failed

### ISSUE THREAD

**Issue by oldmylei** [espressif/esp-hosted#407]
Hello, I tried to connect our board to the ESP32 development board, and there was an error when loading the driver.

The following is the diagram of the hardware connection. We connected seven signals with the ESP32 in total, and one GND signal was not drawn:
![d40ba2d336354f3c4eb316b072365bd](https://github.com/espressif/esp-hosted/assets/126750652/3250baa9-5cc9-4e19-9902-52617accf62e)

When I finished this, I built and burned the ESP32 board image under esp_hosted_ng/esp/esp_driver/network_adapter, and compiled the driver of our board under esp_hosted_ng/host. Below is the log of the ESP32C3 development board and the dmesg content information of our board:
[esp32_dmesg.txt](https://github.com/user-attachments/files/15922495/esp32_dmesg.txt)
[esp32_log.txt](https://github.com/user-attachments/files/15922498/esp32_log.txt)

According to the log, it seems that my board is not properly wired to the ESP32 development board? Or the ESP32 needs to start some kind of mode.

Could you take a look, please? Thank you

**Comment by mantriyogesh** [comment]
Thank you for sharing detailed logs and clear issue description @oldmylei .

Currently ESP driver code not able to claim spi bus.
Meaning possibly you are using incorrect spi bus and/or incorrect cs.
Other possibility that spidev(default Linux user space driver for spi) is still using the spi bus. If the bus is available only then hosted kernel driver can claim.. if spidev is still enabled, (check `$ ls /dev/spidevX.Y` output. X is bus, Y is CS. If bus-cs mapping file is present, spidev has claimed it. You might need to disable spidev in device tree or SoC config)

Please review carefully, https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/docs/porting_guide.md

**Comment by oldmylei** [comment]
![image](https://github.com/espressif/esp-hosted/assets/126750652/2872c12d-9ae2-4605-9f17-2a084e94f65d)
I read this tutorial and it required me to turn off the default SPI driver. Here is the SPI node configuration I used. I actually disabled it and went to the device node that did not generate /dev/spidevX.Y

![image](https://github.com/espressif/esp-hosted/assets/126750652/5fe5b92e-7098-467c-8b37-38bfae64746b)

**Comment by mantriyogesh** [comment]
Although we are not experts for device tree of your Linux SoC, From experience, It looks like you are disabling spi bus altogether. you need to only disable spidev (disable the user space driver only. not the bus itself)

**Comment by oldmylei** [comment]
![image](https://github.com/espressif/esp-hosted/assets/126750652/eb78b4a9-48a2-442f-aaf6-422f56700768)
![image](https://github.com/espressif/esp-hosted/assets/126750652/3e1ac874-3de4-451a-bbcd-99eda699a825)

I think I should understand what you mean. Do you mean to mask only the default driver of SPI as I did in the picture above?

**Comment by mantriyogesh** [comment]
Change status disabled for (inside) **spidev**. not outside it.
Outside would disable 'spi6' altogether.

I am not sure of compatible. But I think, let it be as is for now.

**Comment by mantriyogesh** [comment]
You would need to port the spi bus number to 6 and cs to 0(?).
Also connect the smallest wires possible with spi6 and CS0 of spi 6 with slave' connections.

There are three extra wires/GPIOs apart from 4 wire spi, that you need.
Please refer porting guide in earlier comment.

Once those handshake, dataready and resetpin are configured, test them.
reset pin on toggle should reset slave.
data_ready and handshake are interrupt at host, so manually low to high (edge) interrupt should either print in hosted code (check porting guide)

**Comment by mantriyogesh** [comment]
these are typical steps while porting..

**Comment by oldmylei** [comment]
I set handshake, dataready and resetpin as normal output GPIO
![image](https://github.com/espressif/esp-hosted/assets/126750652/d5db1728-4356-4753-b1d9-0f3966beaca9)

IO90 is handshake, then IO91 is dataready, I didn't connect the resetpin signal because the ESP32C3 has a reset button on the development board
After doing this, I still get the following error when loading the compiled generated driver

![image](https://github.com/espressif/esp-hosted/assets/126750652/61e063b3-a54a-4a3b-a048-21bf4770deae)

**Comment by oldmylei** [comment]
![image](https://github.com/espressif/esp-hosted/assets/126750652/fc4bcf9d-962e-4781-b4d8-d34ab49c33c6)
Good morning, according to the documentation, it seems that only 4.X kernel adaptation is described. My system kernel is version 5.10, can it adapt to this module?

**Comment by mantriyogesh** [comment]
HCI over SPI is supported 'over' kernel version 4.7.0.
Meaning, 4.7.0 + is supported for HIC over SPI (bluetooth over spi).

You kernel version is supported.

**Comment by mantriyogesh** [comment]
> I didn't connect the resetpin signal because the ESP32C3 has a reset button on the development board

Reset not only a buttin, but GPIO on devkit header, named 'RST'

Can you please run command,
```
ls /dev/spidev*
```
-> 1. Without disabling spidev
-> 2. After disabling spidev

Also can you please post the current device tree for spi?

**Comment by oldmylei** [comment]
good morning, I noticed that under esp-hosted-release-ng-v1.0.2/esp_hosted_ng/host there isa spidev_disabler.dts file, the following is its contents
![image](https://github.com/espressif/esp-hosted/assets/126750652/f678dbe7-5c4d-40b9-a4f2-6fbe03d80430)
Maybe I should add the following information to my device tree?
![image](https://github.com/espressif/esp-hosted/assets/126750652/a3844f69-d836-4eeb-90f8-647a1ab7117c)

**Comment by oldmylei** [comment]
![image](https://github.com/espressif/esp-hosted/assets/126750652/449ffbd1-a639-4a57-91cc-196c86758676)
When my device tree is set this way, my machine generates spidev6.0 nodes
![image](https://github.com/espressif/esp-hosted/assets/126750652/a5537e43-6249-4dc8-beac-8097abc1e546)

![image](https://github.com/espressif/esp-hosted/assets/126750652/e24ff002-b96f-4d74-9b2e-defb5832e8d4)
When my device tree is set this way, no spi nodes are generated

**Comment by oldmylei** [comment]
In addition, I set the reset pin to 89, which proved to be effective, the ESP32 development board reset when I entered the following command
![image](https://github.com/espressif/esp-hosted/assets/126750652/dee2da62-4c40-4f99-83e6-fcb4662f577e)
![image](https://github.com/espressif/esp-hosted/assets/126750652/a2651e84-f82c-4ca4-a604-2fab88f2838c)

**Comment by oldmylei** [comment]
Hello, I have solved the problem of loading the driver, but I did not generate the WIFI node after loading the esp32_spi.ko driver, the following is the detailed log

root@myd-jd9x:~# insmod esp32_spi.ko resetpin=89
root@myd-jd9x:~#
root@myd-jd9x:~#
root@myd-jd9x:~# lsmod
Module                  Size  Used by    Tainted: G
esp32_spi              49152  0
lt8912b_hdmi           16384  0
yt6801                368640  0
xrp_hw_semidrive       16384  0
lm75                   20480  0
pvrsrvkm             1273856  0
ngbe                  217088  0
root@myd-jd9x:~#
root@myd-jd9x:~#
root@myd-jd9x:~# ifconfig
lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:168 errors:0 dropped:0 overruns:0 frame:0
          TX packets:168 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:12416 (12.1 KiB)  TX bytes:12416 (12.1 KiB)

root@myd-jd9x:~# ifconfig -a
can0      Link encap:UNSPEC  HWaddr 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00
          NOARP  MTU:16  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:10
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
          Interrupt:52

can1      Link encap:UNSPEC  HWaddr 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00
          NOARP  MTU:16  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:10
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
          Interrupt:53

can2      Link encap:UNSPEC  HWaddr 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00
          NOARP  MTU:16  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:10
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
          Interrupt:54

can3      Link encap:UNSPEC  HWaddr 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00
          NOARP  MTU:16  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:10
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
          Interrupt:55

enp1s0    Link encap:Ethernet  HWaddr 00:55:7B:B5:7D:F7
          BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
          Interrupt:50 Base address:0x8000

eth0      Link encap:Ethernet  HWaddr 56:B8:E5:23:5E:F3
          BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
          Interrupt:46

eth1      Link encap:Ethernet  HWaddr 76:AF:C2:A6:A3:16
          BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
          Interrupt:47

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:168 errors:0 dropped:0 overruns:0 frame:0
          TX packets:168 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:12416 (12.1 KiB)  TX bytes:12416 (12.1 KiB)

teql0     Link encap:UNSPEC  HWaddr 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00
          NOARP  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:100
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

tunl0     Link encap:UNSPEC  HWaddr 00-00-00-00-30-30-00-30-00-00-00-00-00-00-00-00
          NOARP  MTU:1480  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

root@myd-jd9x:~#

**Comment by oldmylei** [comment]
This is the info information for loading the driver
![aacdd0373e2d40b5af8a5e0bf437fdf](https://github.com/espressif/esp-hosted/assets/126750652/51d035b1-9c34-49d6-b703-f2cfd1cd393e)

**Comment by mantriyogesh** [comment]
ifconfig listing is later stage. Also, have you changed the name of network intterface as such? (by default, it would be espsta0 for ng)
I think something is wrong before reaching to ifconfig stage. dmesg and esp side logs would let you know the issues.

**Comment by mantriyogesh** [comment]
As resetpin is not hooked, please reset esp manually and check the logs in the dmesg again..

Reset pin is mandatory.

For now,
1. rmmod esp32_spi
2. insmod esp32_spi resetpin=89
3. check esp is being reset on step (2)
4. also check the GPIO handshake and data ready correctly generating interrupts. 
For this, Can add log at https://github.com/espressif/esp-hosted/blob/8db84e42e71823e6a04ed55eef06d3ccacd8e4f6/esp_hosted_ng/host/spi/esp_spi.c#L60
and https://github.com/espressif/esp-hosted/blob/8db84e42e71823e6a04ed55eef06d3ccacd8e4f6/esp_hosted_ng/host/spi/esp_spi.c#L69

and manually trigger posedge interrupt, to see these pins are correct.

**Comment by mantriyogesh** [comment]
Also , please test **spidev** test (before disabling spidev), to verify the underlying spi driver is working fine or not for your bus and chip select combination.

**Comment by mantriyogesh** [comment]
In summary, do the porting (using porting guide).
Verify all GPIOs work as they are intended to be working.
Verify SPI lines are working fine (both hardware and software/config issues would need tto rule out first)

**Comment by oldmylei** [comment]
All I know for sure is that the resetpin I set will reset the ESP32C3 properly

**Comment by mantriyogesh** [comment]
sure. other pins working as they expected to be?

Unless all GPIOs work as they need to, porting is incomplete and solution would not work as such

**Comment by oldmylei** [comment]
![image](https://github.com/espressif/esp-hosted/assets/126750652/55e9263c-7ba6-4411-8b95-471bde8e6ac0)
I added the above printing information, the result did not print these sentences

![image](https://github.com/espressif/esp-hosted/assets/126750652/52121d37-70bb-44cb-ad27-6da2a2de00ce)
Is it proof that my handshake and dateready pins are not connected correctly, I am sure that I have set both pins to GPIO output mode and are GPIO90 and GPIO91 pins

**Comment by oldmylei** [comment]
I noticed the spi/esp_spi.h file, do I need to change the HANDSHAKE_PIN and SPI_DATA_READY_PIN values?
![image](https://github.com/espressif/esp-hosted/assets/126750652/a617f3f8-65ab-4659-9648-17a97ec979ba)

**Comment by mantriyogesh** [comment]
> https://github.com/espressif/esp-hosted/issues/407#issuecomment-2196405152

This is mentioned in porting guide..
![Screenshot 2024-06-28 at 4 31 44 PM](https://github.com/espressif/esp-hosted/assets/67477199/0507f0fc-7ee1-47ab-bbf7-9592b0c8dc8f)

**Comment by oldmylei** [comment]
![image](https://github.com/espressif/esp-hosted/assets/126750652/c6d84b7c-7a48-4a32-838d-c45cd823ee3a)
After I changed the values of HANDSHAKE_PIN and SPI_DATA_READY_PIN, an interrupt did occur and the debugging information above, but still no wifi node was generated

**Comment by mantriyogesh** [comment]
Worth to check the porting guide step-wise.
Porting is prerequisite for ESP-Hosted project. We have added common issues and ways to fix them in this guide.

**Comment by oldmylei** [comment]
Good morning. I have a question. Can the CS signal of SPI be set to GPIO to control it?

**Comment by mantriyogesh** [comment]
Sorry, are you referring to the CS at slave (on ESP32-C6) ?
If you mean can you change the CS to different GPIO, yes, you can do it.

Or do you mean you wish to omit it from SPI slave controller and manually handle it?

**Comment by oldmylei** [comment]
I added the following print information to the esp_inetaddr_event function in the esp_cfg80211.c file
![image](https://github.com/espressif/esp-hosted/assets/126750652/65d435c8-86c4-423e-8345-0586b0ac146e)
But the log information is not printed, is there some reason why this function is not called?

**Comment by oldmylei** [comment]
In addition, I traced this function to the end, which was called by the esp_add_card function, and then the esp_add_card function was called by the process_event_esp_bootup function in the spi/esp_spi.c file, specifically why this function was not called. How do we track it?

**Comment by mantriyogesh** [comment]
@kapilkedawat ++

**Comment by kapilkedawat** [comment]
@oldmylei as Yogesh mentioned, please check the porting guide and make sure connections are set correctly. Only after that we need check further for cfg functions. Unless connections are not correct you won't get the boot event and cfg functions won't be called.

**Comment by mantriyogesh** [comment]
@oldmylei I think instead of jumping directly to end solution, as we know it is not working, can set back a little and progress step wise. 

Assuming, pins resetpin, handshake and data ready are doing good, we can focus on verifying the transport instead of complicating with network and wifi layers. 

Have you tested spidev testing before on the spi line, possibly with full duplex? What I meant, did we verify spi line in either direction?

In any way, Can you perform raw throughout test in either direction (tx and rx) using [transport_test](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/docs/Raw_TP_Testing.md).

If this test fails, above stated spidev test would be mandatory. If this test passes, please attach the logs for ESP & host for both directions (rx and tx separately)

**Comment by oldmylei** [comment]
![image](https://github.com/espressif/esp-hosted/assets/126750652/226f724b-fb69-4bc3-b5c1-fa75948b4282)
Good morning, I have written an spi test program, and the above is the verification result (I short-connected the two feet MISO and MOSI for verification), and the result shows that the received data is normal

![image](https://github.com/espressif/esp-hosted/assets/126750652/d3815fc0-8e6c-4955-8f77-6741b7d912d5)
In addition, I would like to ask whether I need to change the parameter information of esp_board.chip_select if my CS chip selects the signal using GPIO control

**Comment by mantriyogesh** [comment]
We have never tried in Linux. As part MCU we used to control when the default supported CS was not correct.
But as you are using Linux, It is better to let the SPI controller driver handle this own their own.

However, if you think you must do this, It would be your own debugging and handling the timings correct. Logic analyser would help you in getting there in big way.

Is there any specific rationale behind that you could share with us why you wish to handle it manually?

If you wish to control the pin manually, you need to find a way to inform your linux driver, **not to use** chip_select. Also, modify lower linux driver or if supported through config, disable handling of chip select.

**Comment by mantriyogesh** [comment]
Also worth tto check if your platform SPI driver supports `SPI_CS_HIGH` mode

**Comment by oldmylei** [comment]
![image](https://github.com/espressif/esp-hosted/assets/126750652/7e10c1c4-e109-4fbe-806e-dd538e0f4ad8)
In fact, there is a notification driver on my underlying driver, which is the pin of my CS chip selection signal

**Comment by mantriyogesh** [comment]
I did not understand your latest comment.

In any case, this is all experiments you wish to do it.
As scope of ESP-Hosted, it can only function if the base spi transport is stable.

So I would request to test your manual changes with some other simple user space driver first in either direction rx and tx.

Once you verify the spi driver is stable , you can add ESP-Hosted layer on top to satisfy your use case.

**Comment by oldmylei** [comment]
hi,mantriyogesh
I have now switched to the SDIO+UART interface on the ESP32,However, after I connected the hardware (a total of eight signals, namely SD2.CLK CMD DATA0~DATA3 resetpin and GND) and loaded the driver, there was no data interactive display. The printed result is as follows

![Uploading image.png…]()

**Comment by oldmylei** [comment]
hi,mantriyogesh
I have now switched to the SDIO+UART interface on the ESP32,However, after I connected the hardware (a total of eight signals, namely SD2.CLK CMD DATA0~DATA3 resetpin and GND) and loaded the driver, there was no data interactive display. The printed result is as follows

![image](https://github.com/espressif/esp-hosted/assets/126750652/2e660140-bfe5-4a44-a7e8-67ea464d20be)

**Comment by mantriyogesh** [comment]
First only try with sdio, instead of sdio+uart.

As uart pins are not connected, you cannot expect uart to be working. Please check again setup document in detail.

Also share the setup picture for sdio.

**Comment by mantriyogesh** [comment]
Troubleshoot your issues at first level using [porting guide](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/docs/porting_guide.md)

Once you keep every requirement fulfilled in porting guide, sdio would work.

**Comment by oldmylei** [comment]
I am following the diagram below to connect the signal

![image](https://github.com/espressif/esp-hosted/assets/126750652/42f21d39-b0ab-4774-96f8-d1c85a907ac7)

Also, I made progress when loading the driver last night, but reported the following error
![0a80a462b4cdfa40c0abe36485eefeb](https://github.com/espressif/esp-hosted/assets/126750652/8da72e8b-70fd-475f-a6bf-167858f079ce)

I want to change mmc1 to mmc2, here is my configuration in the device tree, but I can't find the source code where to configure this

![image](https://github.com/espressif/esp-hosted/assets/126750652/f17ce0a9-07a4-4338-a30a-398d4da1de6b)

**Comment by mantriyogesh** [comment]
1. -110 is timeout.
Please share dmesg from bootup.

Additionally,
2. Can you please try 'broken-cd;'

3. Have you added necessary pull-ups to necessary lines?
4. Can you please share the photo of your setup?

**Comment by oldmylei** [comment]
[log.txt](https://github.com/user-attachments/files/16154319/log.txt)
Above is dmesg from bootup.

The second point I don't understand is the need to add 'broken-cd; 'Is this parameter on my device tree? Are there any examples?

For the third point, I connected the other six SD2 signals except CLK to pull-up resistors

Fourth, I don't quite understand what you mean. Do you need me to share anything

**Comment by mantriyogesh** [comment]
1. Possible incorrect/unsupported Device Tree blob present in your Device tree (Just mild warning)
```
[    7.124899] mmcblk mmc1:0001: no of_node; not parsing pinctrl DT
```
seems to have some incorrect config. But I would rather prefer to **ignore** if the mmc device instance is tried to probe by kernel fine (looks like so in your case)

2. ESP sdio device is sensed by your kernel, so basic connections should be fine (still need to judge connections for signal integrity issues discussed in the end of this comment)
![Screenshot 2024-07-10 at 10 53 55 AM](https://github.com/espressif/esp-hosted/assets/67477199/d09f5756-9077-449e-9d07-91d67286ea83)

3. Please make sure you connect **resetpin** to 'EN' of ESP32 as mentioned in porting guide:
![Screenshot 2024-07-10 at 10 55 29 AM](https://github.com/espressif/esp-hosted/assets/67477199/459f98dd-a6bd-4065-9771-d724f9972d0a)

4. Entry, with 'mc1:0001:**2**' failing can be ignored, since we do not use `function 2` of sdio. Although, 'mc1:0001:1' is failing with `-110`. Try polling using 'broken-cd'
```
[  277.643854] esp_sdio: probe of mmc1:0001:2 failed with error -110
```

> The second point I don't understand is the need to add 'broken-cd; 'Is this parameter on my device tree? Are there any examples?

5. yes, This is to poll the probing, if it fails to probe slave sdio.
you can find example from https://wiki.stmicroelectronics.cn/stm32mpu/index.php?title=SDMMC_device_tree_configuration&oldid=70703.


> For the third point, I connected the other six SD2 signals except CLK to pull-up resistors

6. What is pull-up value for all sdio pins?  Recommendation in [porting_guide](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/docs/porting_guide.md#13-additional-pull-ups)
![Screenshot 2024-07-10 at 10 47 12 AM](https://github.com/espressif/esp-hosted/assets/67477199/60e75328-32fb-47dc-9fdc-ba41722fd5de)

7. Please cross-check if you need to burn eFuse for voltage expectations.
For ESP32, most likely you would need to do this irreversible action. It highly depends on exact module/chipset used. Please check : [Bootstrap conflict on SDIO dat2](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/sd_pullup_requirements.html#conflicts-between-bootstrap-and-sdio-on-dat2)

> Fourth, I don't quite understand what you mean. Do you need me to share anything

8. **Jumper cables & SDIO**
Is your hardware PCB or you are using wires? SDIO is known for signal integrity issues if used with wires. If you wish to evaluate with wires only, there are some suggestions:
- shorten wires as much as possible
- if you can do 2-3 cm, great
- Good quality wires only
- Use exact same length for all the pins.

9. 1-bit mode SDIO
If still wish to evaluate using jumper cables, can use 1-bit SDIO mode. In 1-bit mode, dat0 and dat1 are used. Still other unused pins dat2 and dat3 need to be pulled-up.
In this case also point (8) is applicable. No changes need to be done at ESP. For Linux, device tree need to be changed with 'bus-width=<1>' and re-flash the device tree  & reboot.

### Some general suggestions while using Device Tree (DT)
- Beware and very conservative while adding any known or unknown config in your device tree.
- Do not introduce too many changes at a time and test, you wouldn't know what line makes the issues
- Although it is generic config across Linux systems, Not all Linux system driver implement all device tree options
- At times, the low level Linux drivers are also buggy or not thought of interoperability for any specific config

**Comment by oldmylei** [comment]
I think the point is, I want to use mmc2 to process this signal from SDIO, but it is clearly recognized by mmc1, how can I modify this

**Comment by mantriyogesh** [comment]
This is not in the scope of ESP-Hosted.
Getting the platform up with device tree or config as per your Linux, is **prerequisite** for ESP-Hosted project as such.

But we understand that you are stuck till getting platform up, to evaluate ESP-Hosted as such.
So please consider this as help and not support as such.

1. Backup existing device tree (source and built dtb)
2. Identify the procedure to recover your device if 'boot-loop' happens while executing below steps
3. Find dtsi for your board. DSI you can consider as reference-book for your Linux all possible hardwares
4. Create small dts with pointing the 'mmc2' device blob as reference and update the config similar to mmc1
5. build with new dts (make sure it is visible to your build environment) and flash the new dtb to specific place needed as per your Linux SoC.

Please bear in mind, Espressif is not responsible should you encounter any hardware failures or bootloops or any data loss.

**Comment by oldmylei** [comment]
![lQLPKdGuDc8VaFPNBGXNBOmw_X_eXbB8wpcGewsUiJOKAA_1257_1125](https://github.com/user-attachments/assets/fa5a8adc-72a1-4683-9663-40ec5b4e80c5)
Hello, now the driver is loaded at this stage, but there seems to be a problem with the communication, can you help to check

**Comment by mantriyogesh** [comment]
Please help by sharing a photo of your setup.

**Comment by oldmylei** [comment]
![image](https://github.com/user-attachments/assets/61ff2254-5e75-4fa5-8860-38211eb9f12d)

**Comment by mantriyogesh** [comment]
Picture of where ESP is connected to Linux.

**Comment by oldmylei** [comment]
![lQDPJxYcixK8KUPNBQDNA8CwRYo9Ln7izr0Gex1f_dFmAA_960_1280](https://github.com/user-attachments/assets/71dd9902-1930-4268-9907-1947caad0294)

**Comment by oldmylei** [comment]
![image](https://github.com/user-attachments/assets/f95d1c2b-e384-48a5-ba4e-6fe7e6559b47)
I have connected pull-up resistors for several signals of SD2 in the part I circled in the figure, and do you think the DuPont wire we connected is too long

**Comment by mantriyogesh** [comment]
Yes, Wires should at all not be used.
If you use wires, the communication issues might happen.

If you can reduce to **3-4 cm**, a little better!!
Still, occasional signal integrity issues might be there.

- I highly suggest: point (9) SDIO 1-bit mode for **evaluation**  
  - Less data pins = less noise = less signal integrity issues
- Fallback to SPI for wires                                                             OR
- You have base communication already tested, create **PCB** with SDIO connections and then use 4-bit SDIO mode.

### Things to keep in mind, while creating PCB for SDIO
- You keep the points for tapping all SDIO signals
- Keep extra GPIOs (maybe floating/NC) in between ESP and host, for some future use
- **Ensure lengths of SDIO are not very high from host to slave**
- Pull-ups added with 51KOhm for expected lines
- Keep exact same length of connections in PCB for all the SDIO lines.

**Comment by oldmylei** [comment]
Good evening, I have cut the length of Dupont wire, the following is the error appeared in ESP32, can you see
![be7260d3bbdcb4719e01c5a5e474d870](https://github.com/user-attachments/assets/b8e1ddea-ff54-441b-864c-b35bf0b90a81)

**Comment by oldmylei** [comment]
![1f3d4bba61fad1abd5e2cd71d3f8fb43](https://github.com/user-attachments/assets/df09ddb5-bfc1-4aa2-8f2e-7b5338f49b24)
When executing the above command, the above ESP32 error will appear

**Comment by oldmylei** [comment]
![b96ba3f48930194d872fb230b8901791](https://github.com/user-attachments/assets/bb610869-9ffc-4997-bb58-f323013b4754)

**Comment by mantriyogesh** [comment]
Check your power to ESP board or gpios one by one
https://esp32.com/viewtopic.php?t=6469

**Comment by oldmylei** [comment]
![1f3d4bba61fad1abd5e2cd71d3f8fb43](https://github.com/user-attachments/assets/8e2d2054-b2d2-499a-802c-f4aba3888cfe)
If I do not execute this command, the ESP32 module will not crash for a long time, I think it should have nothing to do with power supply

**Comment by mantriyogesh** [comment]
The command you fire, has wifi real work involved.
Unless you fire wifi, wifi has no action.

Once wifi library, it increases power consumption.

Have you burnt the eFuse?

**Comment by oldmylei** [comment]
![image](https://github.com/user-attachments/assets/bed96ad7-5799-487e-ade6-f864517eb374)
It looks like I can't program it into the ESP32 right now, and I can't erase it

![image](https://github.com/user-attachments/assets/6e752e92-eb39-4e61-944d-405ab96f2ca5)
Connect to the ESP32 serial port, and the print looks like this

**Comment by oldmylei** [comment]
When I used wpa_supplicant command to open wifi, I encountered some problems, the following is the error message
[wpa_wifi_log.txt](https://github.com/user-attachments/files/16201215/wpa_wifi_log.txt)

**Comment by oldmylei** [comment]
[wpa_wifi_log.txt](https://github.com/user-attachments/files/16201249/wpa_wifi_log.txt)
After  run the wpa_supplicant -D nl80211 -i espsta0 -c /root/open.conf command, the ESP32 crashes

![image](https://github.com/user-attachments/assets/fd7f091f-4527-4c1c-832e-5d85a764302e)

**Comment by mantriyogesh** [comment]
send full ESP log. textual logs only, no photos.
Share from the beginning.

Also share dmesg from system bootup. Again no photos, only textual logs.

**Comment by oldmylei** [comment]
ok ,this is log
[ESP32_log.txt](https://github.com/user-attachments/files/16201308/ESP32_log.txt)
[wifi_error.log](https://github.com/user-attachments/files/16201310/wifi_error.log)

**Comment by mantriyogesh** [comment]
How are you powering ESP in https://github.com/espressif/esp-hosted/issues/407#issuecomment-2225378827?

**Comment by mantriyogesh** [comment]
I did not get answer ifor question asked in https://github.com/espressif/esp-hosted/issues/407#issuecomment-2225452016

**Comment by oldmylei** [comment]
> How are you powering ESP in [#407 (comment)](https://github.com/espressif/esp-hosted/issues/407#issuecomment-2225378827)?

I used the USB cable to power the ESP32 module, as shown in the image below

![image](https://github.com/user-attachments/assets/755b4786-72b2-4dfa-a2ce-70944dec04ef)

**Comment by oldmylei** [comment]
> The command you fire, has wifi real work involved. Unless you fire wifi, wifi has no action.
> 
> Once wifi library, it increases power consumption.
> 
> Have you burnt the eFuse?

I don't quite understand this reply, I can attach the device tree configuration of my current SD2

![image](https://github.com/user-attachments/assets/241afb41-d72c-4ef2-977a-806c342465c7)

**Comment by mantriyogesh** [comment]
Did you burnt the eFuse?

I had pointed a very detailed reply in https://github.com/espressif/esp-hosted/issues/407#issuecomment-2219459154.

Please check each step and let me know step wise.

Step (7) is something related to eFuse.

**Comment by oldmylei** [comment]
If I set bus-width = <1>, this setting will only use dat0 and dat1, not dat2. In addition, when I type python espefuse.py set_flash_voltage 3.3V on my board, the following log appears.
[espefuse.log](https://github.com/user-attachments/files/16229295/espefuse.log)

**Comment by mantriyogesh** [comment]
I had gone through your dmesg. And I find OOM triggered in your log (which is serious). I don't know how much memory and stuff you got running over it. he crash shows:

![Screenshot 2024-07-15 at 9 58 02 AM](https://github.com/user-attachments/assets/bf1d40f4-531a-475e-b0fe-d116f3d7193a)

I fail to understand how these symbols are related to esp-hosted.
These are not generated from esp-hosted.
All I understand is it possibly generated through ntpd, 
![Screenshot 2024-07-15 at 10 16 47 AM](https://github.com/user-attachments/assets/26b2c2bd-5679-41e6-9a90-c847bdf7f5e5)

Not sure, if it is the root cause or an after-effect. I am not sure, but it would be good to review processes listed while oom was triggered in your dmesg log.
If you can show how it is triggered from ESP-Hosted, we can still have a look.

### Unrelated (or related?) suggestion
It's just friendly advice, nothing specific to esp-hosted as such. I see a lot of device tree configuration errors are present in your config, which are complained in dmesg. Can you ignore, can you not? I am not really not sure.
![Screenshot 2024-07-15 at 10 22 51 AM](https://github.com/user-attachments/assets/be6a5e92-c3a7-4006-ad79-48c9ba3c890e)

I think errors such would create some peripherals work incorrect way. Worst case, some invalid config and GPIOs not treated as errors will interfere, the correct working of other peripherals. If had been in your shoe, I would have enabled only necessary GPIOs and only necessary peripherals, so that no warning, error or no unintentional interference in between working of different peripherals.

Regarding your error, obviously it is serious. But it is worth to find out which kernel module or which exact code is generating such. Next question would be, do you need that module or peripheral or hardware/software.

Again, To be honest, I am not an expert, but If not something not needed/used, I would prefer to disable specific hardware/software. In fact I would selectively enable only the hardware/software I wish to work with. To disable any hardware from your device tree, instead of changing in original dtsi, its best to refer the blob in separate dts and just change the status to disabled.

System stability is very important, I suggest you to review your device tree and config by self or someone expert at it.

**Comment by oldmylei** [comment]
Thanks, but I was wondering if it was necessary to execute the espefuse.py tool when I set bus-width = <1>, because you said it would only use the dat0 and dat1 bits of SD2

**Comment by mantriyogesh** [comment]
### What to do with DAT2 if 1-bit SDIO to be used?

Actually, the answer lies in the earlier link I had shared.
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/sd_pullup_requirements.html#conflicts-between-bootstrap-and-sdio-on-dat2

Scroll down and check point (2) under `Conflicts Between Bootstrap and SDIO on DAT2`

Here it is discussed, what to do with DAT2 GPIO, if it is 1-bit mode that you wish to enable.

**Please read the documentation** and then come back here.
Regarding config at slave, We have tested that slave auto understands that sdio 1-bit mode is needed and no change is required.
But optionally, if you wish to let slave know upfront, that **do not** even release 4-bit sdio mode, you can add
```
config.flags |= SDIO_SLAVE_FLAG_DAT2_DISABLED;
```
at line, 
https://github.com/espressif/esp-hosted/blob/91b093365cc58214532b976dd8749f65523d1e6d/esp_hosted_ng/esp/esp_driver/network_adapter/main/sdio_slave_api.c#L131

And re-flash the esp, which would make sure that slave releases SDIO in 1-bit mode **only**.
As the document says, host **must** use 1-bit SDIO in this case (I assume, your current device tree is already flashed with `bus-width=<1>` and rebooted the host)

### Reverting to 4-bit SDIO later
- For 4-bitt mode to be used, you should first remove above `SDIO_SLAVE_FLAG_DAT2_DISABLED` change and re-flash the ESP. 
- Revert device tree to use 4-bit SDIO and flash tthe dtb & reboot.

**Comment by mantriyogesh** [comment]
I had updated/edited above comment if in case missed to notice.

**Comment by oldmylei** [comment]
![image](https://github.com/user-attachments/assets/e2090974-accb-41ba-93fc-ad360865a588)
https://github.com/espressif/esp-hosted/issues/407#issuecomment-2227629703

After I set bus-width = <1>, I used the above method to make the firmware burned to the ESP32, the problem still occurs
https://github.com/espressif/esp-hosted/issues/407#issuecomment-2226807690

**Comment by mantriyogesh** [comment]
@oldmylei , this issue of `TG0WDT_SYS_RESET` has nothing related with esp-hosted as such.

1. Please cross-check all your connections from esp to host.
https://github.com/espressif/arduino-esp32/issues/9137
2. If you did any soldering, please cross-check if there is proper resistance in between soldered adjacent pins and not short.
3. Can you replace the Laptop to ESP USB-A-to-microUSB cable, to smaller one?
4. Is the ground connected fine? Can you please add more ground lines in between esp and host, it stabilises the signal.
5. Is ESP accidentally touching any other pin on host, for example, ground, etc?
\
If all above did not work,
6. `rm -rf sdkconfig build` and rebuild-reflash esp32
7. Your current idf is on 4.4. Change IDF to **latest stable** branch,
check how to [install esp-idf](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/get-started/index.html) & install latest stable esp-idf branch (v5.2.2 as of now), repeat step (6)

**Comment by myLiuWeiQin** [comment]
At present, we use 1-bit SD mode, which can generate espsta0 network interface. However, when we perform WiFi connection operation, ESP32 module crashes. I think it has nothing to do with hardware connection, because ESP32 crashes every time when we perform WIFI connection operation. Can you analyze the cause of the crash according to the crash log?

**Comment by mantriyogesh** [comment]
Have you even read the full comment?
Please check https://github.com/espressif/esp-hosted/issues/407#issuecomment-2226824088 you attached.
This is power issue either at ESP or GPIO.

Please get back once you check all points.

**Comment by oldmylei** [comment]
How to Obtain the esp-idf branch (v5.2.2 as of now)

**Comment by mantriyogesh** [comment]
Please refer https://github.com/espressif/esp-hosted/issues/407#issuecomment-2229942770, there is hyperlink already provided.

Did you test steps (1) to (6)?

**Comment by myLiuWeiQin** [comment]
For the test procedure you provided：
（1）：The link you provided is about SPI connection, but we have carefully checked SDIO connection according to the figure below
![1721112054927](https://github.com/user-attachments/assets/cde12fdf-7b0b-44ef-81ef-a0f22775951b)

（2）：We made the SDIO connection according to the figure below, and added a ground connection. At the same time, we also checked the various signals, and there was no short circuit between the signals

![808dc4230b5c1a07b8adb0924597e27](https://github.com/user-attachments/assets/87f83af5-b7d4-42d2-a492-d6db51349f52)

（3）：According to your analysis, it is believed that the problem is power supply, so we use direct power supply instead of HUB power supply to obtain stable power supply

![4f9ddf613371f9db546d62bea3eb311](https://github.com/user-attachments/assets/a7836252-7ad0-4528-9ef5-a561619c8da1)

Next we are going to try rm -rf sdkconfig build and install the esp-idf branch of v5.2.2

**Comment by oldmylei** [comment]
https://github.com/espressif/esp-hosted/issues/407#issuecomment-2229942770
![image](https://github.com/user-attachments/assets/4a0f6e1a-67c3-4cc7-a616-b26de95c08c3)

As for the seventh point, we have been using the ESP32 burning firmware provided by you, rather than compiling it ourselves

**Comment by mantriyogesh** [comment]
Host and slave needs to be on the same commit. Different firmware commit and different hosted commit leads to undefined behaviour

**Comment by myLiuWeiQin** [comment]
We are according to you provide a link (https://github.com/espressif/esp-hosted/releases/tag/release%2Fng-v1.0.2), compile the driver, but there was no ESP the source firmware, Therefore, we cannot confirm whether the host and slave are in the same commit

**Comment by mantriyogesh** [comment]
Checkout master, build ESP , build host and use them.

**Comment by myLiuWeiQin** [comment]
The link you provided does not contain ESP, how to build it

**Comment by mantriyogesh** [comment]
Please check documentation, https://github.com/espressif/esp-hosted/tree/master/esp_hosted_ng

**Comment by oldmylei** [comment]
![image](https://github.com/user-attachments/assets/49225956-ed64-4216-84d1-d4121df4471c)
First of all have to confirm about (https://github.com/espressif/esp-hosted/releases/tag/release%2Fng-v1.0.2) links the content inside
This link does not have the ESP32 source code, only the ESP32 burning firmware, followed by the driver source.
If you need to modify the firmware of ESP32, do you need to give us a corresponding source code?

**Comment by mantriyogesh** [comment]
Please check documentation at https://github.com/espressif/esp-hosted/issues/407#issuecomment-2230437549

Full source code and procedure to flash is provided.

**Comment by oldmylei** [comment]
Now that we have set bus-width = <4> and burn efuse, the espsta0 node will not disappear, but the following situation will occur.

[bord_log.txt](https://github.com/user-attachments/files/16263521/bord_log.txt)
[ESP32_log.txt](https://github.com/user-attachments/files/16263522/ESP32_log.txt)
[dmesg_log.txt](https://github.com/user-attachments/files/16263540/dmesg_log.txt)
@mantriyogesh

**Comment by oldmylei** [comment]
Looks like he's been trying to reconnect

**Comment by mantriyogesh** [comment]
Thank you, @oldmylei, for providing all detailed logs.

## Two issues from logs
- [1] AP connection request received, AP not connected.
- [2] Transport error - timeout

While [1] is something we check from the logs, [2] is serious (but even expected on current hardware arrangement)

### [1] AP connection request received, AP not connected.
@kapilkedawat++ Can you PTAL?
reason I see: WIFI_REASON_HANDSHAKE_TIMEOUT (204)

### [2] Transport error - timeout
@oldmylei 
I still see :
```
[  189.643315] esp32: wait_and_decode_cmd_resp(priv, cmd_node) failure, ret: -1
```

#### 2.1 Check transport only speed
Which makes me ask you to run [transport only test](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/docs/Raw_TP_Testing.md) in both rx and tx directions one by one.
Let us know frequency you use and the throughput in either direction you get.

#### 2.2 Final solution
- You would need finally the PCB to run over SDIO with 4-bit mode (no choice).
- [Keep in mind while creating 4-bit mode PCB](https://github.com/espressif/esp-hosted/issues/407#issuecomment-2225140423)

#### 2.3 Current issue debugging
For easy alternative only for evaluation, would to re-run with
A. lower frequency
B. 1-bit mode
C. both (A) and (B)

**Comment by oldmylei** [comment]
One thing I want to confirm is that the esp-hosted/esp_hosted_ng/host driver generates a "wlan0" wifi node, which is different from the espsta0 node in your document and the espsta0 node in the rpi_int.sh script.
I just want to make sure I have the correct version of this driver

![image](https://github.com/user-attachments/assets/6b6e8dc8-b413-40b6-a5df-0c2931ddfe92)
![image](https://github.com/user-attachments/assets/c4bd4626-c850-4506-b5d2-8e4bc47c8b91)

**Comment by mantriyogesh** [comment]
Can you please post output of
```
ifconfig -a
```

**Comment by oldmylei** [comment]
sure

[wlan0_log.txt](https://github.com/user-attachments/files/16279234/wlan0_log.txt)

**Comment by mantriyogesh** [comment]
Yes looks to be issue, we will fix this asap.

**Comment by mantriyogesh** [comment]
(espsta0) naming issue resolved, please use latest master - commit da21e362d5bced2e8ed9bc4eadb9cffb1d6c1be1

**Comment by oldmylei** [comment]
Since we need to use AP mode, we decided to use fg to compile the firmware and drivers, but we ran into the following problems. Here's the log.

[bord_log.txt](https://github.com/user-attachments/files/16281415/bord_log.txt)
[dmesg_log(1).txt](https://github.com/user-attachments/files/16281417/dmesg_log.1.txt)
[ESP32_log.txt](https://github.com/user-attachments/files/16281418/ESP32_log.txt)

**Comment by oldmylei** [comment]
After I set the random hardware addr, I reported an error when assigning AP mode to the node. Below is the log
[AP_log.txt](https://github.com/user-attachments/files/16281936/AP_log.txt)

**Comment by mantriyogesh** [comment]
> https://github.com/espressif/esp-hosted/issues/407#issuecomment-2236128035

1. You cannot run 50Mhz on your wired device! Running 5Mhz should work fine with wired setup.
Once evaluation is done, get PCB and then bump the frequency.
2. I clearly see brownout being detected. Please check this happens.
Provide sufficient input power to ESP, otherwise, it would fail in undefined way. (luckily, brownout would take the chip down by the time)
- **Change your power cable** , with smaller one, which supports some decent voltage and ampere rating.

3. I see you are running FG SDIO+UART.
Please be informed
- A. you **cannot** run **FG firmware** with **NG kernel module** and vice versa.
- B. Use exact **same commit** at esp and host. Always use **current master** at both sides.
- C. If you are evaluating SDIO, **do not flash/configure UART**, use the `HCI over SDIO`, instead of HCI over UART.
- D. For FG, you need to port and run `rpi_init.sh` fully (check documentation), also [porting_guide](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/Linux_based_host/porting_guide.md)
- D. Just inserting a module is not going to get FG working. Please check the procedure to run [demo app](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/common/c_demo.md)
- E. **Do not skip the documentation**. Every documentation step you skip, you would just delay the working.

**Comment by mantriyogesh** [comment]
> https://github.com/espressif/esp-hosted/issues/407#issuecomment-2236204330

This is wrong.
![Screenshot 2024-07-18 at 7 08 19 PM](https://github.com/user-attachments/assets/c32b8bdb-f5d9-4433-999a-8da81fd414d6)

Follow [documentation](https://github.com/espressif/esp-hosted) for [FG](https://github.com/espressif/esp-hosted/tree/master/esp_hosted_fg).

**Comment by oldmylei** [comment]
![5677480a3eaa1e588e7a1000383dc382](https://github.com/user-attachments/assets/b9184f4b-0b2e-4fb5-8a37-c6b53b791673)
I tried to compile this demo application of FG, but some libraries and intermediate files are missing

**Comment by mantriyogesh** [comment]
Use the correct cross compile and pass it to make command while building demo app.

Alternatively, https://github.com/espressif/esp-hosted/blob/da21e362d5bced2e8ed9bc4eadb9cffb1d6c1be1/esp_hosted_fg/host/linux/host_control/c_support/Makefile#L1-L9 could be changed as per your 'target' building environment and trigger `make`

You can use same arguments you used while building the kernel module.
Check [cross-compile](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/Linux_based_host/porting_guide.md#23-cross-compilation) documentation.

**Comment by myLiuWeiQin** [comment]
![image](https://github.com/user-attachments/assets/a9214c12-40a4-4bf8-ac67-049a88be1938)
We specified the path to the compiler, but it didn't work

**Comment by mantriyogesh** [comment]
[which files are crtl or crtn.c ](https://askubuntu.com/questions/75116/how-do-you-build-older-gcc-versions-errors-about-crti-o-not-being-found)

The failures here are typical for cross-compilation.

To simply further, just cross-compile any simple program,
```c
$ cat test_app.c

#include <stdio.h>
void main ()
{
	printf("Hello!!!\n");
}
```
and load in your target. once you able to get it working, then build this app again. Another issue is pthread lib is not visible. but this you can check from your sdk. worst case, provide static library to make it work.

Check your SoC documentation, they generally provide cross-compilation steps to build the sample examples.

**Comment by myLiuWeiQin** [comment]
After modifying your Makefile, I successfully compiled test.out, but I still couldn't set the mac address
![image](https://github.com/user-attachments/assets/37d4fda5-0502-49c7-b26c-018aefa23ca6)
![image](https://github.com/user-attachments/assets/b53817da-46a7-47f5-8cac-fe4d49fd7946)
![image](https://github.com/user-attachments/assets/d70af585-d41c-4e30-b63b-9d894bc39e76)

**Comment by mantriyogesh** [comment]
Check [control_apis](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/common/ctrl_apis.md) supported.

softap_start is the api you are looking. documentation covers in https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/common/c_demo.md#list-of-control-path-commands

complete demo code is at: https://github.com/espressif/esp-hosted/tree/master/esp_hosted_fg/host/linux/host_control/c_support

check test.c main() as start point.

**Comment by mantriyogesh** [comment]
Again, check doc completely in link https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/common/c_demo.md#some-points-to-note.

**Comment by myLiuWeiQin** [comment]
I directly set the mac address and the printing message said that it was successful, but the actual setting was not successful. In addition, I set the WiFi mode to softap mode, indicating that the setting was successful, but the query told me that it was station mode now. This makes me feel strange, is there a documentation for the whole networking process? Or you can test the reliability of this tool

**Comment by mantriyogesh** [comment]
> I directly set the mac address and the printing message said that it was successful, but the actual setting was not successful

This is (incorrect) assumption that you manually the mac address and it should work. Demo app would get mac from esp and set it for you. in [demo app doc](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/common/c_demo.md) earlier shared, it is written:
![Screenshot 2024-07-19 at 5 21 49 PM](https://github.com/user-attachments/assets/f51fc035-d219-47d0-9e1d-dec72459aeff)


> In addition, I set the WiFi mode to softap mode, indicating that the setting was successful, but the query told me that it was station mode now

Demo app is just a primitive way of showcasing the control APIs. Unless you check the small short code in test.c, I think you cannot imagine. There is nowhere mentioned (or otherwise as well) that get_mode accepts argument.
Further, 

Please understand, the demo app is kept smallest and simplistic and separate from kernel module.
It is way to showcase the real code on top , how the control apis are to be invoked, not built much smart, as anyway expected by developers to be replaced.

If you wish, you can try python code, which is tested against a lot of checks etc. it supports many ways to interact with control apis.
https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/common/python_demo.md
python: 
![image](https://github.com/user-attachments/assets/3fbc12a6-091d-44e4-824f-85970ed44e88)

**Comment by myLiuWeiQin** [comment]
Ok, we successfully connected to the AP node of ESP32, thank you for your support

**Comment by oldmylei** [comment]
At present, there is still a problem, that is, the file ctrl_config.h needs to be modified every time the SSID and password of wifi are set. Is there any way to do this at the application layer

![image](https://github.com/user-attachments/assets/18a3b953-37a7-44b1-8917-49dd1e9ad436)

**Comment by mantriyogesh** [comment]
Use python, it works in command line mode, with flexible arguments.
These are demo applications, just ways to showcase control apis to get working.

Ensure that the lower layer is not changed. Anything, above [control_apis](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/common/ctrl_apis.md) functions, could be modified as you wish.
